alternator/conditions.cc

```diff
@@ -131,6 +132,37 @@ static bool check_BEGINS_WITH(const rjson::value* v1, const rjson::value& v2) {
     return val1.substr(0, val2.size()) == val2;
 }
 
+// Check if a JSON-encoded value equals any element of an array.
+static bool check_IN(const rjson::value* val, const rjson::value& array) {
+    if (array.Size() < 1) {
+        throw api_error("ValidationException", "IN operator requires nonempty AttributeValueList");
+    }
+    if (!array[0].IsObject() || array[0].MemberCount() != 1) {
+        throw api_error("ValidationException",
+                        "IN operator requires AttributeValueList elements to be of type String, Number, or Binary ");
+    }
+    const auto& type = array[0].MemberBegin()->name;
+    if (type != "S" && type != "N" && type != "B") {
+        throw api_error("ValidationException",
+                        "IN operator requires AttributeValueList elements to be of type String, Number, or Binary ");
```

Comment:
Please add a test that confirms that non-scalars are not allowed in IN - I don't think we have one, right?
A test, which you can run against DynamoDB, is a great way to see that what you think you should be doing - is actually what DynamoDB does.